### DISCUSSION TITLE
Play Framework for a large application (ERP)

### DISCUSSION THREAD

**Discussion by Knitter** [playframework/playframework#12349] [General]
Hi,

I'm currently evaluating options to move a large application (+80 controllers, +300 actions that either return HTML based on template system or return JSON, +400 database tables on various schemas). Does anyone have any experience with large, enterprise applications made with Play that could share some insights?

Some of these controllers will be converted and isolated into services, but most of the application will be classic MVC, supported by a PostgreSQL database; if using Play, then scala templates will be used for most of the UI with some screens being converted to AJAX/JavaScript UI (mostly vanilla, no UI framework).

At the moment, the two options that are being considered are Play (no decision yet if Scala or Java) and  Spring; I've already used Play with Java in the past, and recently picked Play 3 and Scala for some integration services that were needed, but not enough to understand if Play can support such application (think ERP).

Thanks

**Comment by irumiha** [comment]
I once worked on an application of a comparable size. It had its problems but I don't think Play was one of them. Assuming you follow general application layout best practices, at that scale the amount of code that will interact with Play directly should be less than 10% of the overall codebase.

**Comment by jducoeur** [comment]
In my experience, Play can deal with that level of volume and complexity,
but you'll need to get the details right. In particular, it's crucial to
make sure that your code is written to be non-blocking and asynchronous (so
you don't wind up thread-bound), and pay attention to stuff like
serialization (a bad serializer can hurt performance when you are dealing
with high volume).

**Comment by Knitter** [comment]
Hi,

Thank you for your comments, I expect Play to handle the type of application I'm working on, but most of what we can find as examples is either too small and focused on a single feature or too complex and not relatable to our use case.

Currently, the existing services are already asynchronous in all tasks that would block (mostly file access and/or database) and I intend to keep doing that in the larger application.

Thanks.

**Reply by mkurz** [reply]
@Knitter did you decide already? I do not see any reason why you would not be able to migrate your application to Play, be it Java or Scala.
Regarding sync/async, you can find more information here: https://www.playframework.com/documentation/3.0.x/ThreadPools (actually you could also write your application in a highly synchronous fashion, where you have many threads, just like "classic" web apps, see the "Highly synchronous" section).
We do have sponsors who do use Play in large setups, be it Java or Scala, so I am very confident Play would not _the_ problem. Also, it comes with many features you mostly can configure/enable/(de)activate/ in `application.conf`, thanks to Play's plugin system, which also means you can extend it yourself quite easily based on dependency injection.
Let me know if you need any guidances for a specific use case and/or problem.

**Reply by Knitter** [reply]
Hi,

We're in the process of testing both Play and Spring, it will take us some time to do as we've chosen a few screens/features of our current application and implement them in Play (Scala) and  Spring and, due to some other internal reasons, I've also decided that each team member will repeat the projects. So we'll each implement the same (very small) set of features in both Play and Spring and then compare/decide. Personally, I would like to go with Play/Scala, as I've had some experience with Play in the past, I like the framework and also don't think that could be the problem, but I want the rest of the team to test it themselves.
